### PR TITLE
Customizable key glyphs

### PR DESCRIPTION
**Pull request by clauswilke**
This PR implements customizable legend key glyphs, as discussed in #3145.

``` r
library(ggplot2)

base <- ggplot(data.frame(x = 1:3, y = 3:1), aes(x, y))

cowplot::plot_grid(
  base + geom_line(aes(color = "default")),
  base + geom_line(aes(color = "modified"), key_glyph = "timeseries"),
  base + geom_line(aes(color = "modified"), key_glyph = "vpath"),
  base + geom_line(aes(color = "modified"), key_glyph = "abline")
)
```

<sup>Created on 2019-04-11 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

@hadley @thomasp85 @yutannihilation I think this is the simplest and minimally invasive implementation. However, it does some shuffling around with the `params` parameter in `layer()`, so I'd like to hear your perspective. The alternative would require modifications in every geom, and would not automatically work in extension packages until they modify their geoms also.

### CODE REVIEW

**Comment by clauswilke** [comment]
I think this is ready for review now.

**Comment by yutannihilation** [comment]
> The alternative would require modifications in every geom, and would not automatically work in extension packages until they modify their geoms also.

Do you think there are actual use cases of `key_glyph` argument? If we don't modify the existing geoms, to me, it seems we'll stick with `params` only.

File: NEWS.md [hunk 1]
```diff
@@ -35,6 +35,9 @@ core developer team.
 * `geom_rug()` gains an "outside" option to allow for moving the rug tassels to 
   outside the plot area (@njtierney, #3085) and a `length` option to allow for 
   changing the length of the rug lines (@daniel-wells, #3109). 
+  
+* All geoms now take a `key_glyph` paramter that allows users to customize
+  how legend keys are drawn (@clauswilke, #3145).
```
Comment (yutannihilation) [inline NEWS.md]: Can we credit @mitchelloharawild for inventing the timeseries glyph here?

File: R/legend-draw.r [hunk 2]
```diff
@@ -200,10 +219,34 @@ draw_key_label <- function(data, params, size) {
...
+      col = alpha(data$colour %||% data$fill %||% "black", data$alpha),
+      lwd = (data$size %||% 0.5) * .pt,
+      lty = data$linetype %||% 1,
       lineend = "butt"
     )
   )
 }
+
+
+
```
Comment (yutannihilation) [inline R/legend-draw.r]: Seems too much blank lines.

File: NEWS.md (hunk 1 above)
Comment (clauswilke) [reply NEWS.md]: Done.

File: R/legend-draw.r (hunk 2 above)
Comment (clauswilke) [reply R/legend-draw.r]: Fixed.

**Comment by clauswilke** [comment]
> Do you think there are actual use cases of `key_glyph` argument? If we don't modify the existing geoms, to me, it seems we'll stick with `params` only.

I added this to avoid a secret, undocumented parameter. I'm not sure how to handle this best. One perspective is that all geoms should define a `key_glyph` argument and then pass it on to `layer()` explicitly. However, then setting this argument to `NULL` in the geom constructors would look weird, and setting it to the actual key glyph in the geom constructor would always override the setting in the `Geom`, which is also weird. The way I did it seemed the best compromise. Sufficiently hidden that it's not in your face, but sufficiently present that it is discoverable.

**Comment by yutannihilation** [comment]
As for documentation, I think we can list and describe notable parameters in the doc of `params`, but I too am not sure what's the best way for this... I want to wait for others' comments.

**Comment by thomasp85** [comment]
>Do you think there are actual use cases of key_glyph argument? If we don't modify the existing geoms, to me, it seems we'll stick with params only.

I confess to not really understanding the above. @yutannihilation can you elaborate?

**Comment by clauswilke** [comment]
@thomasp85 What @yutannihilation was referring to is that now `layer()` has a `key_glyph` argument but there is currently no code that actually uses it, because the geoms provide `key_glyph` via the `params` argument. He asked whether we should eliminate the `key_glyph` argument from `layer()`. My opinion is "no", and I explained it here: https://github.com/tidyverse/ggplot2/pull/3236#issuecomment-482763195

**Comment by thomasp85** [comment]
Ah, I see. Yeah, I like your approach

**Pull request merged**